nativelib/src/main/scala/scala/scalanative/posix/pthread.scala

```diff
@@ -0,0 +1,303 @@
+package scala.scalanative
+package posix
+
+import scala.scalanative.native.{
+  CFunctionPtr0,
+  CFunctionPtr1,
+  CInt,
+  CSize,
+  CUnsignedInt,
+  Ptr,
+  extern,
+  name
+}
+import scala.scalanative.posix.sched.sched_param
+import scala.scalanative.posix.sys.types._
+import scala.scalanative.posix.time.timespec
+
+// SUSv2 version is used for compatibility
+// see http://pubs.opengroup.org/onlinepubs/007908799/xsh/threads.html
+
+@extern
+object pthread {
+
+  def pthread_atfork(prepare: routine, parent: routine, child: routine): CInt =
+    extern
+
+  def pthread_attr_destroy(attr: Ptr[pthread_attr_t]): CInt = extern
+
+  def pthread_attr_getdetachstate(attr: Ptr[pthread_attr_t],
+                                  detachstate: Ptr[CInt]): CInt = extern
+
+  def pthread_attr_getguardsize(attr: Ptr[pthread_attr_t],
+                                guardsize: Ptr[CSize]): CInt = extern
+
+  def pthread_attr_getinheritsched(attr: Ptr[pthread_attr_t],
+                                   inheritsched: Ptr[CInt]): CInt = extern
+
+  def pthread_attr_getschedparam(attr: Ptr[pthread_attr_t],
+                                 param: Ptr[sched_param]): CInt = extern
+
+  def pthread_attr_getschedpolicy(attr: Ptr[pthread_attr_t],
+                                  policy: Ptr[CInt]): CInt = extern
+
+  def pthread_attr_getscope(attr: Ptr[pthread_attr_t],
+                            scope: Ptr[CInt]): CInt = extern
+
+  def pthread_attr_getstacksize(attr: Ptr[pthread_attr_t],
+                                stacksize: Ptr[CSize]): CInt = extern
+
+  def pthread_attr_init(attr: Ptr[pthread_attr_t]): CInt = extern
+
+  def pthread_attr_setdetachstate(attr: Ptr[pthread_attr_t],
+                                  detachstate: CInt): CInt = extern
+
+  def pthread_attr_setguardsize(attr: Ptr[pthread_attr_t],
+                                guardsize: CSize): CInt = extern
+
+  def pthread_attr_setinheritsched(attr: Ptr[pthread_attr_t],
+                                   inheritsched: CInt): CInt = extern
+
+  def pthread_attr_setschedparam(attr: Ptr[pthread_attr_t],
+                                 param: Ptr[sched_param]): CInt = extern
+
+  def pthread_attr_setschedpolicy(attr: Ptr[pthread_attr_t],
+                                  policy: CInt): CInt = extern
+
+  def pthread_attr_setscope(attr: Ptr[pthread_attr_t], scope: CInt): CInt =
+    extern
+
+  def pthread_attr_setstackaddr(attr: Ptr[pthread_attr_t],
+                                stackaddr: Ptr[Byte]): CInt = extern
+
+  def pthread_attr_setstacksize(attr: Ptr[pthread_attr_t],
+                                stacksize: CSize): CInt = extern
+
+  def pthread_cancel(thread: pthread_t): CInt = extern
+
+  def pthread_cleanup_push(routine: CFunctionPtr1[Ptr[Byte], Unit],
+                           arg: Ptr[Byte]): Unit = extern
+
+  def pthread_cleanup_pop(execute: CInt): Unit = extern
+
```

Comment:
`pthread_cleanup_push` and `pthread_cleanup_pop` are implemented as macros, so you can't use them—not directly at least.

This pair of calls is basically a try-finally control structure for ensuring proper resource cleanup when one thread kills another. Cancelling threads is not a very common practice these days—instead, the common practice is at most to let one thread "interrupt" another, with the interrupted thread then having a chance to decide whether it wants to exit in an orderly fashion or resume its previous business.

Conceivably a workaround should be possible, but I doubt it would justify the effort.